gtk4/src/subclass/entry_buffer.rs

```diff
@@ -0,0 +1,256 @@
+                )
+            }
+        }
+    }
+}
+
+unsafe impl<T: EntryBufferImpl> IsSubclassable<T> for EntryBuffer {
+    fn override_vfuncs(class: &mut glib::Class<Self>) {
+        <Object as IsSubclassable<T>>::override_vfuncs(class);
+
+        let klass = class.as_mut();
+        klass.delete_text = Some(entry_buffer_delete_text::<T>);
+        klass.deleted_text = Some(entry_buffer_deleted_text::<T>);
+        klass.get_length = Some(entry_buffer_get_length::<T>);
+        klass.get_text = Some(entry_buffer_get_text::<T>);
+        klass.insert_text = Some(entry_buffer_insert_text::<T>);
+        klass.inserted_text = Some(entry_buffer_inserted_text::<T>);
+    }
+}
+
+unsafe extern "C" fn entry_buffer_delete_text<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+    position: u32,
+    n_chars: u32,
+) -> u32 {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+
+    imp.delete_text(wrap.unsafe_cast_ref(), position, n_chars)
+}
+
+unsafe extern "C" fn entry_buffer_deleted_text<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+    position: u32,
+    n_chars: u32,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+
+    imp.deleted_text(wrap.unsafe_cast_ref(), position, n_chars)
+}
+
+static GET_TEXT_QUARK: Lazy<glib::Quark> =
+    Lazy::new(|| glib::Quark::from_string("gtk4-rs-subclass-entry-buffer-text"));
+
+unsafe extern "C" fn entry_buffer_get_text<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+    n_bytes: *mut usize,
+) -> *const libc::c_char {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+
+    let ret = imp.get_text(wrap.unsafe_cast_ref());
+    *n_bytes = ret.len();
+    // Ensures that the returned text stays alive for as long as
+    // the entry buffer instance
+    let fullptr = ret.to_glib_full();
+    wrap.set_qdata(*GET_TEXT_QUARK, fullptr);
+    fullptr
+}
+
+unsafe extern "C" fn entry_buffer_get_length<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+) -> u32 {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+
+    imp.get_length(wrap.unsafe_cast_ref())
+}
+
+unsafe extern "C" fn entry_buffer_insert_text<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+    position: u32,
+    charsptr: *const libc::c_char,
+    n_chars: u32,
+) -> u32 {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+    let text: Borrowed<GString> = from_glib_borrow(charsptr);
+
+    let pos_end = text
+        .char_indices()
+        .map(|(i, _char)| i)
+        .nth(n_chars as usize)
+        .unwrap();
```

Comment:
I'm not very sure about the unwrap here, should we assume that the returned text's characters length > n_chars here?

Reply:
If you check the docs, `n_chars` can also be `-1`. So that needs to be handled at least and means "all characters".

Otherwise I don't know, seems like a bug if there are fewer characters available so panic?

Reply:
Also check if any of the other n_bytes/n_chars can be -1 please

Reply:
They all do except the n_bytes one, I wonder how the delete_text|deleted_text should be handled ? I have also a question regarding this one: it seems the generated code uses an unsigned integer, can I just cast as a signed one to check if it's negative?

Reply:
Yes or check against `u32::MAX` (that's the same as `-1i32`).

For these cases I usually make it an `Option<u32>` on the Rust side and make the `-1` to `None`.